set -eo pipefail

curl -sL https://storage.googleapis.com/gcp-public-data--gnomad/release/3.1.1/vcf/genomes/gnomad.genomes.v3.1.1.sites.$1.vcf.bgz | bcftools annotate --threads 2  -x '^INFO/AF_non_cancer,^INFO/AF_non_cancer_afr,^INFO/AF_non_cancer_ami,^INFO/AF_non_cancer_asj,^INFO/AF_non_cancer_eas,^INFO/AF_non_cancer_fin,^INFO/AF_non_cancer_mid,^INFO/AF_non_cancer_nfe,^INFO/AF_non_cancer_oth,^INFO/AF_non_cancer_raw,^INFO/AF_non_cancer_sas,^INFO/AF_non_cancer_amr,^INFO/AC,^INFO/AN,^INFO/AF,^INFO/nhomalt,^INFO/AC_popmax,^INFO/AN_popmax,^INFO/AF_popmax,^INFO/nhomalt_popmax,^INFO/AC_controls_and_biobanks,^INFO/AN_controls_and_biobanks,^INFO/AF_controls_and_biobanks,^INFO/AF_non_cancer,^INFO/primate_ai_score,^INFO/splice_ai_consequence' | /vt/vt normalize - -n -r Homo_sapiens_assembly38.fasta  | bgzip -@4 -c > gnomad.genomes.v3.1.1.sites.$1.bcftools_INFO_subset.vt_norm.vcf.gz
